src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -268,52 +269,88 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
             var projects = (await solutionManager.GetNuGetProjectsAsync()).OfType<IDependencyGraphProject>().ToList();
             var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
+                await actionBlock.SendAsync(new ProjectRestoreSpec()
```

Comment:
is `context` thread-safe, or rather, does `GetPackageSpecsAndAdditionalMessagesAsync` use `context` in a thread-safe way?

If the action block is changed to `ActionBlock<IDependencyGraphProject>`, you can send the project to it, so `project.GetPackageSpecsAndAdditionalMessagesAsync` also runs in parallel. Right now it runs sequentially. I think that would also remove the need for the additional private class

Reply:
Unfortunately, it looks it's not thread-safe
Also one more observation is if I move `project.GetPackageSpecsAndAdditionalMessagesAsync` move inside my [parallization code ](https://github.com/NuGet/NuGet.Client/pull/3605/commits/09af23be602ef600dc100d9c728e4eee09c9409b) then whole runtime was slow as it was sequential.  But it's before I use TPL, not sure for TPL.
It looks moving async call as first call inside method was not good move.